syntax = "proto3";
package Avalon.Network.Packets.Abstractions;

message NetworkPacket {
  NetworkPacketHeader Header = 1;
  bytes Payload = 2;
}
enum NetworkPacketFlags {
  // this is a composite/flags enumeration
  None = 0;
  ClearText = 1;
  Handshake = 2;
  Encrypted = 4;
}
message NetworkPacketHeader {
  NetworkPacketType Type = 1;
  NetworkPacketFlags Flags = 2;
  NetworkProtocol Protocol = 3;
  int32 Version = 4;
}
enum NetworkPacketType {
  option allow_alias = true;
  UNKNOWN = 0;
  ERROR = -1;
  CMSG_SERVER_INFO = 4096;
  CMSG_CLIENT_INFO = 4097;
  CMSG_CLIENT_HANDSHAKE = 4098;
  CMSG_AUTH = 8192;
  CMSG_AUTH_PATCH = 8193;
  CMSG_LOGOUT = 8194;
  CMSG_REGISTER = 8195;
  CMSG_CHARACTER_LIST = 8208;
  CMSG_CHARACTER_CREATE = 8209;
  CMSG_CHARACTER_DELETE = 8210;
  CMSG_CHARACTER_SELECTED = 8211;
  CMSG_CHARACTER_LOADED = 8212;
  CMSG_MAP_TELEPORT = 8224;
  CMSG_QUEST_STATUS = 8256;
  CMSG_QUEST_LIST = 8257;
  CMSG_QUEST_QUERY_AVAILABLE = 8258;
  CMSG_INTERACT = 8240;
  CMSG_REQUEST_SERVER_VERSION = 8225;
  CMSG_REQUEST_LOBBY_LIST = 8195;
  CMSG_MOVEMENT = 8196;
  CMSG_PING = 8197;
  CMSG_PONG = 8198;
  CMSG_CHAT_MESSAGE = 8199;
  CMSG_CHAT_OPEN = 8200;
  CMSG_CHAT_CLOSE = 8201;
  CMSG_GROUP_INVITE_RESULT = 8202;
  SMSG_SERVER_INFO = 12288;
  SMSG_SERVER_HANDSHAKE = 12290;
  SMSG_SERVER_HANDSHAKE_RESULT = 12289;
  SMSG_AUTH_RESULT = 12304;
  SMSG_LOGOUT = 12305;
  SMSG_REGISTER_RESULT = 12306;
  SMSG_CHARACTER_CONNECTED = 12320;
  SMSG_CHARACTER_DISCONNECTED = 12321;
  SMSG_CHARACTER_LIST = 12323;
  SMSG_CHARACTER_CREATED = 12324;
  SMSG_CHARACTER_DELETED = 12325;
  SMSG_CHARACTER_SELECTED = 12326;
  SMSG_MAP_TELEPORT = 12336;
  SMSG_PONG = 12294;
  SMSG_PING = 12295;
  SMSG_SERVER_VERSION = 12290;
  SMSG_PLAYER_POSITION_UPDATE = 12293;
  SMSG_NPC_UPDATE = 12296;
  SMSG_CHAT_MESSAGE = 12297;
  SMSG_CHAT_OPEN = 12298;
  SMSG_CHAT_CLOSE = 12299;
  SMSG_GROUP_INVITE = 12301;
  SMSG_GROUP_INVITE_RESULT = 12302;
}
enum NetworkProtocol {
  Invalid = 0;
  Tcp = 1;
  Udp = 2;
  Both = 3;
}
